algebird-core/src/main/scala/com/twitter/algebird/CountMinSketch.scala

```diff
@@ -441,23 +460,89 @@ case class CMSZero[K](override val params: CMSParams[K]) extends CMS[K](params)
  */
 case class CMSItem[K](item: K, override val totalCount: Long, override val params: CMSParams[K]) extends CMS[K](params) {
 
-  override def +(x: K, count: Long): CMS[K] = CMSInstance[K](params) + item + (x, count)
+  override def +(x: K, count: Long): CMS[K] = SparseCMS[K](params) + (item, totalCount) + (x, count)
 
   override def ++(other: CMS[K]): CMS[K] = {
     other match {
       case other: CMSZero[_] => this
-      case other: CMSItem[K] => CMSInstance[K](params) + item + other.item
-      case other: CMSInstance[K] => other + item
+      case other: CMSItem[K] => CMSInstance[K](params) + (item, totalCount) + (other.item, other.totalCount)
+      case _ => other + item
     }
   }
 
-  override def frequency(x: K): Approximate[Long] = if (item == x) Approximate.exact(1L) else Approximate.exact(0L)
+  override def frequency(x: K): Approximate[Long] = if (item == x) Approximate.exact(totalCount) else Approximate.exact(0L)
 
-  override def innerProduct(other: CMS[K]): Approximate[Long] = other.frequency(item)
+  override def innerProduct(other: CMS[K]): Approximate[Long] = Approximate.exact(totalCount) * other.frequency(item)
 
 }
 
 /**
+ * A sparse Count-Min sketch structure, used for situations where the key is highly skewed.
+ */
+case class SparseCMS[K](exactCountTable: Map[K, Long],
+  override val totalCount: Long,
+  override val params: CMSParams[K]) extends CMS[K](params) {
+  import SparseCMS._
+
+  override def +(x: K, count: Long): CMS[K] = {
+    val currentCount = exactCountTable.getOrElse(x, 0L)
+    val newTable = exactCountTable.updated(x, currentCount + count)
+    if (newTable.size < maxExactCount) {
+      // still sparse
+      SparseCMS(newTable, totalCount = totalCount + count, params = params)
+    } else {
+      toDense(newTable, params)
+    }
+  }
+
+  override def ++(other: CMS[K]): CMS[K] = {
+    other match {
+      case other: CMSZero[_] => this
+      case other: CMSItem[K] => this + (other.item, other.totalCount)
+      case other: SparseCMS[K] =>
+        // This SparseCMS's maxExactCount is used, so ++ is not communitive
+        val newTable = exactCountTable ++ other.exactCountTable
```

Comment:
wait, we don't want `++` here right? We want `Semigroup.plus(exactCountTable, other.exactCountTable)`.

If that's correct, can you write a test that exposes this bug so we don't have a regression on this?